loki/transform/transform_extract_contained_procedures.py

```diff
@@ -0,0 +1,198 @@
+
+    # Produce derived types appearing in `vars_to_resolve` or in `inner.spec` that need to be resolved
+    # from imports of `procedure`.
+    dtype_imports_to_add = tuple(v.type.dtype for v in vars_to_resolve + inner_spec_vars \
+        if isinstance(v.type.dtype, DerivedType))
+
+    # Produce kinds appearing in `vars_to_resolve` or in `inner.spec` that need to be resolved
+    # from imports of `procedure`.
+    kind_imports_to_add = tuple(v.type.kind for v in vars_to_resolve + inner_spec_vars \
+        if v.type.kind and v.type.kind.scope == procedure)
+
+    # Produce all imports to add.
+    # Here the imports are also tidied to only import what is strictly necessary, and with single
+    # USE statements for each module.
+    imports_to_add = []
+    to_lookup_from_imports = dtype_imports_to_add + kind_imports_to_add + var_imports_to_add
+    for val in to_lookup_from_imports:
+        imp = procedure.import_map[val.name]
+        matching_import = tuple(i for i in imports_to_add if i.module == imp.module)
+        if matching_import:
+            # Have already encountered module name, modify existing.
+            matching_import = matching_import[0]
+            imports_to_add.remove(matching_import)
+            newimport = matching_import.clone(symbols = tuple(set(matching_import.symbols + imp.symbols)))
+        else:
+            # Have not encountered the module yet, add new one.
+            newsyms = tuple(s for s in imp.symbols if s.name == val.name)
+            newimport = imp.clone(symbols = newsyms)
+        imports_to_add.append(newimport)
+
+    ## MAKING THE CHANGES TO `inner`
+    # Change `inner` to take `vars_to_resolve` as dummy arguments and add all necessary imports.
+    # After these lines, `inner` should be self-contained or there is a bug.
+    inner.arguments += tuple(vars_to_resolve)
+    inner.variables += tuple(vars_to_resolve)
+    inner.spec.prepend(imports_to_add)
+
+    ## TRANSFORMING CALLS TO `inner` in `procedure`.
+    # Construct transformation map to modify all calls / function invocations to `inner` in the parent to
+    # include `vars_to_resolve`.
+    # Here any dimensions in the variables are dropped, since they should not appear in the call.
+    # Functions need to be treated differently than Subroutines, hence the "mapping object" `m`.
+    m = {
+        'call_finder': FindInlineCalls() if inner.is_function else FindNodes(CallStatement),
+        'argname': 'parameters' if inner.is_function else 'arguments',
+        'transformer': SubstituteExpressions if inner.is_function else Transformer
+    }
+    call_transmap = {}
+    all_calls = m['call_finder'].visit(procedure.body)
+    inner_calls = (call for call in all_calls if call.routine == inner)
+    for call in inner_calls:
+        newargs = getattr(call, m['argname']) + tuple(map(lambda x: x.clone(dimensions = None), vars_to_resolve))
+        call_transmap[call] = call.clone(**{m['argname']: newargs})
```

Comment:
There are a few issues with this, I think:
* vars_to_resolve variables have scope `inner`, not `procedure`
* If the existing calls have keyword-arguments, this will fail
* While I appreciate the elegance of this, I think the control flow would be much easier to follow if you had separate loop for Subroutines and Functions instead

Without having tested it, something like this might work:

```suggestion
    call_map = {}
    if inner.is_function:
        for call in FindInlineCalls().visit(procedure.body):
            if call.routine is inner:
                newkwargs = tuple((v.name, v.clone(dimensions=None, scope=procedure)) for v in vars_to_resolve)
                call_map = call.clone(kw_parameters=call.kwarguments + newkwargs)
        procedure.body = SubstituteExpressions(call_map).visit(procedure.body)
    else:
        for call in FindNodes(CallStatement).visit(procedure.body):
            if call.routine is inner:
                newkwargs = tuple((v.name, v.clone(dimensions=None, scope=procedure)) for v in vars_to_resolve)
                call_map = call.clone(kwarguments=as_tuple(call.kwarguments) + newkwargs)
    procedure.body = Transformer(call_map).visit(procedure.body)
```

Reply:
Thanks, I had not even thought about keyword arguments. 

After our discussion, we came to the conclusion that indeed it is best to introduce all resolved variables as keyword arguments to calls. This is because if the call already contains keyword arguments, Fortran requires that all arguments that come after them in the call are also keyword arguments. Therefore, introducing resolved variables as positional arguments would require special processing (of the existing call) in case that the call happens to contain keyword arguments.

I changed all tests to check that resolved variables are added to kwargs instead and I added a few new tests for cases with existing arguments present in the call.